storagestate/storagestate.go

```diff
@@ -0,0 +1,730 @@
+ * Public
+ **********************************************************************************************************************/
+
+// New creates storagestate instance.
+func New(cfg *config.Config, messageSender MessageSender, storage Storage) (storageState *StorageState, err error) {
+	log.Debug("Create storagestate")
+
+	storageState = &StorageState{
+		storage:             storage,
+		storageDir:          cfg.StorageDir,
+		stateDir:            cfg.StateDir,
+		messageSender:       messageSender,
+		statesMap:           make(map[aostypes.InstanceIdent]*stateParams),
+		newStateChannel:     make(chan cloudprotocol.NewState, stateChannelSize),
+		stateRequestChannel: make(chan cloudprotocol.StateRequest, stateChannelSize),
+	}
+
+	if err = os.MkdirAll(storageState.storageDir, 0o755); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	if err = os.MkdirAll(storageState.stateDir, 0o755); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	if storageState.watcher, err = fsnotify.NewWatcher(); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	storageMountPoint, err := fs.GetMountPoint(cfg.StorageDir)
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	stateMountPoint, err := fs.GetMountPoint(cfg.StateDir)
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	storageState.isSamePartition = storageMountPoint == stateMountPoint
+
+	go storageState.processWatcher()
+
+	return storageState, nil
+}
+
+// Close close storagestate instance.
+func (storageState *StorageState) Close() {
+	log.Debug("Close storagestate")
+
+	storageState.watcher.Close()
+}
+
+// Setup setup storagestate instance.
+func (storageState *StorageState) Setup(
+	params SetupParams,
+) (storagePath string, statePath string, err error) {
+	storageState.Lock()
+	defer storageState.Unlock()
+
+	instancePath := fmt.Sprintf("%d_%s_%s", params.Instance, params.SubjectID, params.ServiceID)
+
+	log.WithFields(log.Fields{
+		"instance":     params.Instance,
+		"serviceID":    params.ServiceID,
+		"subjectID":    params.SubjectID,
+		"storageQuota": params.StorageQuota,
+		"stateQuota":   params.StateQuota,
+	}).Debug("Setup storage and state")
+
+	storageStateInfo, err := storageState.storage.GetStorageStateInfo(instancePath)
+	if err != nil {
+		if !errors.Is(err, ErrNotExist) {
+			return "", "", aoserrors.Wrap(err)
+		}
+
+		storageStateInfo = StorageStateInstanceInfo{
+			InstancePath:  instancePath,
+			InstanceIdent: params.InstanceIdent,
+		}
+
+		if err = storageState.storage.AddStorageStateInfo(storageStateInfo); err != nil {
+			return "", "", aoserrors.Wrap(err)
+		}
+	}
+
+	if storagePath, err = storageState.prepareStorage(instancePath, params); err != nil {
+		return "", "", aoserrors.Wrap(err)
+	}
+
+	if err := storageState.stopStateWatching(params.InstanceIdent); err != nil {
+		return "", "", aoserrors.Wrap(err)
+	}
+
+	if storageStateInfo.StateQuota != params.StateQuota || storageStateInfo.StorageQuota != params.StorageQuota {
```

Comment:
Better to have separate functions for storage and state.

Reply:
Decided to leave it  as is